[feature] Enter displays wavefront. Double click and F2 rename from the start

atsju opened the pull request:
fix #87 

- [x] press enter displays wft
- [x] F2 renames wft
- [x] ~~right click does not rename anymore~~ Do a real right click menu instead as per George recommandation
- [x] double click has no special effect anymore
- [x] greying out wft names is strange behavior
- [x] create a shortcut for delete wft (dell key)
- [x] any edit will trigger correct signal to reload the wft and have name displayed correctly everywhere
- [x] right click menu can rotate and invert
- [x] fix display bug on delete

I think I achieved something very nice.
Now the actually displayed wft has a small icon in front. I think this is what Dale wanted to achieve with gray/black foreground but it was not 100% working and not always displaying what it should on multi-selection.

Also fixed a bug where wft name could be wrong on metrics window
![image](https://github.com/githubdoe/DFTFringe/assets/4628382/2b59355f-8b47-4f30-9375-d3d449c88f60)

![image](https://github.com/githubdoe/DFTFringe/assets/4628382/7929d36a-ae17-440c-b494-45d42af8590d)

gr5 commented:
I love this but think you should keep the right click functionality for people who expect that.  "delete wft" shortcut should be "Delete" key, right?  Unless you plan to pop up a context menu on right click with "rename" as one of the options.

I am the shortcut key master as I hate using the mouse.  I know probably 1000 shortcuts in windows.  But I didn't know about F2 for rename until I started using Ubuntu as my primary computer.  I was shocked that it works in windows as well.  Now it's the only shortcut I use for renaming because it works in both Windows and Linux.

atsju commented:
I think I'm done with this one.
Lots of small improvement for end user. Little changes for users used to their interface. Probably fixed one or two display bugs in the process.
I refrained from doing any architectural change.

gr5 commented on surfaceanalysistools.cpp:
```diff
@@ -246,24 +288,23 @@ void surfaceAnalysisTools::on_InvertPB_pressed()
-    QModelIndexList indexes = ui->wavefrontList->selectionModel()->selectedIndexes();
-    if (indexes.length() == 1){
-        emit wftNameChanged(indexes[0].row(), NewValue);
+// this is troggered on renaming/editing end
```
troggered -> triggered

atsju replied:
```suggestion
// this is triggered on renaming/editing end
```

gr5 commented on surfaceanalysistools.cpp:
```diff
@@ -110,46 +107,91 @@ void surfaceAnalysisTools::on_spinBox_2_valueChanged(int arg1)
+    contextMenu.addAction(invertAction);
+
+    // Create "Rename" action
+    QAction *renameAction = new QAction("Rename", this);
```
It would be nice if this also showed the shortcut key as F2.  But not mandatory.  Similar for any other actions (delete?) that have a shortcut key.

atsju replied:
Like this ? :

```
Delete (Dell)
Invert
Rename (F2)
Rotate
```

gr5 replied:
![Screenshot from 2023-09-12 16-25-24](https://github.com/githubdoe/DFTFringe/assets/5414808/4f48f974-4374-4c16-acbc-b51b1cea5df8)

"Del" should have one "l" and there must be a way to right justify.  This is how it's done in linux (ubuntu) and alsow windows.  And also Mac.  Right justify the shortcuts.

gr5 replied:
Okay so let's at least do your suggestion with the parenthesis please "()".  And I'll approve.  Or if you can do "right justfied" text than even better!

gr5 replied:
I love it now!  This will help new users quite a bit.

gr5 commented:
Found a bug.  I created a wavefront.  I rotated it.  Second I renamed to ccw.  I then averaged.  Now I have 3 wavefronts.

Now I **deleted** the last one but it's still there I think?  Anyway I did an average of the first two and now I get a blank wft in between where the old "average" was.  When I click on the blank wavefront it is indeed an average just like the one I had deleted.

![Capture](https://github.com/githubdoe/DFTFringe/assets/5414808/3a4776b0-70cb-44d1-9661-16b5e91f2484)

gr5 commented:
I really love the green circle.  

There is a feature where Dale does averaging and if there is a central obstruction, it itakes the obstruction from the first wavefront (I think the one with the green dot as I think it's really the last one clicked maybe if I remember right).

Also the wft that is "first" - the resolution of that wavefront is used for the final average.  Usually the resolution is the same but when you are averaging wavefronts from different people of the same mirror (like the round robin) then it's helpful to be able to pick the resolution.

Anyway, for that reason and many others, I love the green circle.

atsju commented:
> Found a bug. I created a wavefront. I rotated it. Second I renamed to ccw. I then averaged. Now I have 3 wavefronts.
> 
> Now I **deleted** the last one but it's still there I think? Anyway I did an average of the first two and now I get a blank wft in between where the old "average" was. When I click on the blank wavefront it is indeed an average just like the one I had deleted.
> 
> ![Capture](https://user-images.githubusercontent.com/5414808/267454147-3a4776b0-70cb-44d1-9661-16b5e91f2484.PNG)

Oh nice catch. Please reproduce and tell me exactly where to click or select. Some of those bugs are really sensitive to what was selected, clicked,... 
I will go to bed now but will have much time to fix this WE. I will be happy if you find other corner cases. 
Also try multi-selections and selections with arrow keys. 4 eyes will catch more than me alone

gr5 commented:
I'm having trouble duplicating!  Still trying.  But I got distracted because "the bug" happened and FINALLY it happened with your new logging and FINALLY I have a massive clue as to the problem.  Massive.

gr5 commented:
It's not a bug!  I got it to duplicate but it's user error I think.  I think I selected rename but then changed my name and hit "del" to test out the delete shortcut and ended up giving the wavefront a blank name.

atsju commented on surfaceanalysistools.cpp:
```diff
@@ -53,6 +56,8 @@ surfaceAnalysisTools::surfaceAnalysisTools(QWidget *parent) :
     connect(ui->wavefrontList->itemDelegate(), SIGNAL(closeEditor(QWidget*, QAbstractItemDelegate::EndEditHint)), this,
             SLOT(ListWidgetEditEnd(QWidget*, QAbstractItemDelegate::EndEditHint)));
     ui->wavefrontList->setContextMenuPolicy(Qt::CustomContextMenu);
+    QShortcut* dellShortcut = new QShortcut(Qt::Key_Delete, this);
```
```suggestion
    QShortcut* delShortcut = new QShortcut(Qt::Key_Delete, this);
```

atsju commented on surfaceanalysistools.cpp:
```diff
@@ -110,46 +107,91 @@ void surfaceAnalysisTools::on_spinBox_2_valueChanged(int arg1)
- void surfaceAnalysisTools::currentNdxChanged(int ndx){
-     if (ui->wavefrontList->count() == 0)
-         return;
+// this is responsible of givin user feedback about which wft is currently displayed
```
```suggestion
// this is responsible of giving user feedback about which wft is currently displayed
```

atsju commented on surfaceanalysistools.cpp:
```diff
@@ -110,46 +107,93 @@ void surfaceAnalysisTools::on_spinBox_2_valueChanged(int arg1)
+    connect(renameAction, &QAction::triggered, [this, item]() {
+        ui->wavefrontList->editItem(item);
+    });
+    renameAction->setShortcut(QKeySequence(Qt::Key_F2));
```
@Interferometry could you please test this branch on macOS ?
And let me know if you can rename the wft with F2 or only enter ? Then I can put an ifdef for macOS.
Unfortunately there is no QKeySequence::Edit so I need to put the correct key for each platform manually.
https://codebrowser.dev/qt6/qtbase/src/widgets/itemviews/qabstractitemview.cpp.html#2497
@gr5 it can still be merged as is. We can fix this later if needed.

gr5 commented:
I have to admit, the moving green dot confuses me briefly every time because everything is bouncing around left and right.  I mean it's obvious what's going on when you think about it but while it bounces around it seems almost like they are swapping vertically.

I thought I would get used to it but I've been playing with this quite a bit now.

I love the function of the "Enter" key now as well!  I can move around with arrows and hit enter to load that wavefront.  It's really nice.

What do you think of putting a blank space in front of the rows without the green dot?  So they don't shift left/right?

atsju commented:
Might not be cleanest way to do it as it uses a fixed size but it works

gr5 commented:
Using an empty icon to align with the green-circle icon works fine for me.
